Add check for variable/method/class naming

jtraglia opened the pull request:
This check will identify variables/methods/classes with improper naming. For example, if a class does not have an `UpperCamelCase` name or if a constant variable is not `ALL_UPPERCASE`.

A suggested variable name is provided for what I believe the author should use.

Fixes [#4519](https://github.com/ConsenSys/teku/issues/4519) for Teku.

jtraglia committed:
Add check for variable/method/class naming

This BugChecker will identify variables/methods/classes with improper naming.
For example, if a class does not have an UpperCamelCase name or if a constant
variable is not ALL_UPPERCASE.

A suggested fix is provided for what I believe the author might intend.

jtraglia committed:
Rename some functions/variables

jtraglia committed:
All uppercase vars must be static & final

Mistakenly only checked that it was final.

jtraglia commented:
Do you know how to test this branch on another code base, like Teku?

ajsutton commented:
To test you should be able to run `./gradlew publishToMavenLocal` in this repo, then in Teku change the version for errorprone-checks in `versions.gradle` to `develop` ie:
```
dependency 'tech.pegasys.tools.epchecks:errorprone-checks:develop'
```

and teku should then use the version you just built.

I'm getting:
```
/Users/aj/Documents/code/teku/infrastructure/collections/src/main/java/tech/pegasys/teku/infrastructure/collections/cache/NoOpCache.java:36: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
  private NoOpCache() {}
          ^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:

     error-prone version: 2.11.0
     BugPattern: JavaCase
     Stack Trace:
     java.lang.AssertionError
        at com.google.errorprone.fixes.SuggestedFixes.renameMethod(SuggestedFixes.java:759)
        at tech.pegasys.tools.epchecks.JavaCase.matchMethod(JavaCase.java:77)
        at com.google.errorprone.scanner.ErrorProneScanner.processMatchers(ErrorProneScanner.java:449)
        at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:739)
        at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:150)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:898)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
        at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
        at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
        at jdk.compiler/com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
        at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:548)
        at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:150)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
        at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
        at jdk.compiler/com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
        at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:560)
        at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:150)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:591)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
        at com.google.errorprone.scanner.Scanner.scan(Scanner.java:58)
        at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
        at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:152)
        at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1418)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1365)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:960)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:147)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
        at org.gradle.internal.compiler.java.IncrementalCompileTask.call(IncrementalCompileTask.java:89)
        at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:94)
        at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:54)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:39)
        at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkAction.execute(AbstractDaemonCompiler.java:135)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
        at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:97)
        at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
        at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:32)
        at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:22)
        at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:87)
        at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:56)
        at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
        at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
        at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
1 error
```

which suggests you might need to allow constructors to start with a capital. :) And it seems a bit confused with lambdas as well:
```
/Users/aj/Documents/code/teku/infrastructure/exceptions/src/main/java/tech/pegasys/teku/infrastructure/exceptions/ExceptionUtil.java:23: Note: [JavaCase] Use the appropriate case.
      final Throwable err, Class<? extends T> targetType) {
                      ^
  Did you mean 'final Throwable err, Class<? extends T> targetType) {'?
/Users/aj/Documents/code/teku/infrastructure/exceptions/src/main/java/tech/pegasys/teku/infrastructure/exceptions/ExceptionUtil.java:26: Note: [JavaCase] Use the appropriate case.
        .map(e -> (T) e)
             ^
  Did you mean '.map(e -> (T) e)'?
```

ajsutton commented:
But this is really cool - I've been hoping to get these checks in place for quite a while to improve consistency.

jtraglia commented:
Ah thank you! I will fix those issues.

jtraglia committed:
Fix a ton of issues

First of all, fix a bunch of edge cases I hadn't thought of. Some of them:

  * Anonymous classes
  * Test functions
  * Variable names in lambdas

Instead of checking that a variable name meets this specific pattern, check
that the name matches the bare requirements. For example, if a variable name
has starts with an uppercase letter, has a lowercase letter, and does not have
an underscore it is probably an UpperCamel case.

I have tested this against Teku and I think the results look good.

ajsutton commented on src/main/java/tech/pegasys/tools/epchecks/JavaCase.java:
```diff
@@ -0,0 +1,160 @@
+    return Description.NO_MATCH;
+  }
+
+  private static boolean isUnderscore(String name) {
+    return PATTERN_UNDERSCORES.matcher(name).matches();
+  }
```
I'm not sure we need to use a regex for this
```suggestion
  private static boolean isUnderscore(String name) {
    return "_".equals(name);
  }
```

would be faster and more readable.

jtraglia replied:
Ah that regex actually matches one or more underscores. There are often `__` variables in lambdas. I will rename the function to be more clear.

ajsutton commented on src/main/java/tech/pegasys/tools/epchecks/JavaCase.java:
```diff
@@ -0,0 +1,160 @@
+    String name = tree.getName().toString();
+
+    // Class constructors have the <init> name.
+    if (name.equals("<init>")) return Description.NO_MATCH;
```
nit: We generally use braces on all if statements:
```suggestion
    if (name.equals("<init>")) {
      return Description.NO_MATCH;
    }
```

jtraglia replied:
Will update. I totally understand. I like to keep things consistent too.

ajsutton reviewed (commented):
Looking pretty good.  One corner case it doesn't handle yet is fields in interfaces which are inherently `public static final` but don't actually have those flags set (e.g. https://github.com/ConsenSys/teku/blob/b6e15bb7f15a25c2a494e513ae3af18b8dca2c75/infrastructure/ssz/src/main/java/tech/pegasys/teku/infrastructure/ssz/schema/json/SszPrimitiveTypeDefinitions.java#L55)

jtraglia commented:
Ah that is another corner case! I'll try to get it fixed tomorrow morning.

jtraglia committed:
Fix review issues

* Interface variables should be all uppercase.
* Conditionals should have curly braces.
* Renamed function to be more clear.
* Added a few more comments.

jtraglia committed:
Ignore function names for other types of tests

For example, @ParameterizedTest and @TestTemplate.

jtraglia commented:
Okay. I've made those fixes. It wasn't very difficult to do the interface check. Ready for another review.

jtraglia committed:
Add missing curly braces to condition

Oops not sure how I missed those at the bottom of the file.

ajsutton reviewed (approved):
LGTM. Thanks for this.

I'll need to tag a release at some point and then we can pull it into Teku properly but will have to fix or suppress the violations there before we can update (there are classes used with Jackson as part of the REST API that need to use camel cases so the OpenAPI docs come out right).

Probably also worth reviewing your other new rules before doing a release too (I'm running behind today, sorry).

jtraglia commented:
No problem! Thank you for identifying those problems. I'd be happy to fix those case issues in Teku when the time comes. And for those other PRs, no rush. They are relatively complex compared to other checks in the repository.
